### ISSUE TITLE
📐 Building a better combobox (single select)

### ISSUE THREAD

**Issue by substrae** [GCTC-NTGC/gc-digital-talent#5060]
# ⚙️ Summary

### [Visit the Figma file](https://www.figma.com/file/T1R64oNWog10OXiBnuiMkz/Component-library-(GC-Digital-Talent)?node-id=171%3A1557)

@esizer and I feel like we need to better address the combobox (select + keyword) element and its accessibility concerns. Our current third-party component is failing accessibility needs in multiple ways, and so we're setting out to build or find something better that meets the criteria below. Note that this focuses exclusively on a single-select combobox, as we'll expand the component to multi-select at a later time.

## 💾 Files and resources

- N/A

## ✅ Design checklist

- [ ] Component library elements have been used where appropriate
- [ ] Component library elements have been updated where changes where required
- [ ] Newly added interface elements have been added to the component library for reuse
- [ ] Newly added interface elements have been added to Github Discussions using the component issue template
- [ ] The interface is responsive or provides responsive context
- [ ] The interface provides both light and dark mode designs
- [x] Complex interactions have been prototyped in a way that showcases every step in the flow
- [x] Empty/null states have been accounted for
- [x] The design has been reviewed by at least one other team member
- [ ] The design has been presented to the design team
- [ ] The design has been approved by stakeholders

## ✅ Accessibility checklist

**Please refer to [the comprehensive list](https://www.figma.com/file/rRMvm5nq4AMet29PW7bN2p/A11y-annotations?node-id=0%3A1) for more details.**

- [x] Links have meaningful label text
- [x] Navigation matches previously established patterns
- [x] Headings are identified using annotations and follow a logical hierarchy
- [x] Layout is linear and consistent with established patterns
- [ ] New interactive elements have unique focus states that match the established pattern
- [x] Tab order has been identified using annotations
- [ ] Relevant images and icons have appropriate `alt` text
- [x] Use of color meets AA WCAG contrast requirements
- [x] States or information are not identified by color alone
- [x] New interactive elements have appropriately sized touch targets on touch devices

## 📸 Screenshot

![image](https://user-images.githubusercontent.com/6960386/206570941-b8ee63ba-fcf1-4545-a990-76e9d481ad9f.png)

# ✅ Requirements

- [ ] The input should provide a text-based input that triggers a select-style dropdown
- [ ] When the user types in the input, the select's options should return search results for the typed value
- [ ] The user should be able to clear the input using a fixed `x` button
- [ ] The results should accommodate both a "loading" and "no results" state
- [ ] If the user hasn't typed in the input but opens the select options, it should contain ALL available options in alphabetical order
- [ ] The results UI should be navigable using arrow keys, and the `esc` key should close the results and focus the input
- [ ] The search should indicate to assistive tech how many results are being presented, as well as both "loading" and "no results" states
- [ ] The results UI should be limited in height, with a visual affordance indicating that the list is scrollable - this is to help prevent extreme lists of results overlapping critical UI, as well as aiding in responsive navigation of long lists
- [ ] Pressing `enter` while typing in the search should focus the first result in the list

## 🛠️ Clear-able keyword value

Once a value has been typed or selected from the results list, an `x` button should appear next to the caret that when triggered, clears the keywords or selected items.

![image](https://user-images.githubusercontent.com/6960386/206574660-89b0ad50-5db4-4fff-a7e8-cc2b41d9e4a1.png)

## 🛠️ Fixed, scrollable results length

The results list should show between 5-7 results and allow the user to scroll to view more. If more than 5-7 results are a part of the list, the results UI should offer a visual affordance (the shadow) indicating to the user that the list can be scrolled for more results.

![image](https://user-images.githubusercontent.com/6960386/206574724-68e3719f-1046-4bb0-8cbe-02b9bfcb43f1.png)

## 🛠️ Loading state

To account for instances where loading results takes an extended period of time, a loading state should present itself (ideally with an animated icon) to provide immediate feedback that the application is processing the request.

![image](https://user-images.githubusercontent.com/6960386/206574762-5722d5d9-fe5b-4470-b38c-73ed9ce0f9ef.png)

## 🛠️ Alphabetical default results

If the user opens the combobox without typing in a query, they should be provided with a full list of all possible results, ordered in alphabetical order.

![image](https://user-images.githubusercontent.com/6960386/206574803-6c543d30-b4f7-40b7-972c-e9c317e40c19.png)

## 🛠️ No results found state

If the user has provided a keyword or external filter settings that results in no matching results, the results UI should notify them that no results can be found and that they should try different keywords/filters.

![image](https://user-images.githubusercontent.com/6960386/206574853-f55dc08b-f49b-4ea1-9462-d240c0626b0e.png)

### Results ordered by relevance

In an ideal world, the combobox should act like a proper search (particularly for our 140+ skills library), and this means having proper keyword relationships to help surface relevant results. For example, searching "Prototyping" shouldn't just return skills with "Prototyping" in their name, but also "Wireframing", "Interface design", etc.

![image](https://user-images.githubusercontent.com/6960386/206574891-770a8a90-0628-42cb-b14d-bf8db7875bb8.png)

# 🤔 Outstanding concerns

- Multi-select values will be handled later, though the component in Figma does present a possible approach
- There are a variety of accessibility concerns with this interface given its complexity. We should be recording decisions as we make them in reference to both testing and research.

# 📝 Related issues

**Please link all design and development sub-tasks related to this issue here for tracking purposes.**

- #5053

**Comment by mnigh** [comment]
@RM-1978 I noticed you re-opened this issue, is there a reason why? Does it need to have new acceptance criteria? Does it need to be put back through the refinement process?

**Comment by RM-1978** [comment]
Thank @mnigh, @esizer, @substrae and I discussed this issue yesterday. We are exploring ways to make this component more accessible as part of the table work.

For the moment, we are looking at potential ideas to continue updating our approach to this component for accessibility-related reasons. As this particular issue has already been merged, I am completely ok with working with @esizer and @substrae  to create a new issue to update this existing work.

**Comment by substrae** [comment]
Going to just reuse this issue, as a majority of the requirements are remaining the same.

**Comment by mnigh** [comment]
> Going to just reuse this issue, as a majority of the requirements are remaining the same.

@substrae since this issue is intertwined with a PR, would you mind creating a new one instead please?
